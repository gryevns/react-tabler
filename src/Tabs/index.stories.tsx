import React from "react";
import { withKnobs, boolean } from "@storybook/addon-knobs";

import Tabs from ".";

export default {
    title: "Tabs",
    decorators: [withKnobs],
};

export const Default = () => {
    const props = {
        cards: boolean("Card Style", false),
        fill: boolean("Fill Nav", false),
        alt: boolean("Alternative Style", false),
    };
    return (
        <Tabs {...props}>
            <Tabs.Tab title="Home" eventKey="home">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas
                consectetur quam at justo accumsan facilisis. Lorem ipsum dolor sit
                amet, consectetur adipiscing elit. Pellentesque lacinia velit pretium
                nibh pellentesque malesuada. Fusce semper tempor nisi, a pretium
                justo euismod eu. Sed eget euismod nulla. Praesent vitae feugiat
                lacus, quis malesuada nunc. In eleifend posuere ipsum. Pellentesque
                ullamcorper lorem in enim eleifend lacinia. Nam elit quam, venenatis
                in condimentum id, tempor at nisi. Sed congue eu urna sit amet
                iaculis.
            </Tabs.Tab>
            <Tabs.Tab title="About" eventKey="about">
                Quisque in lectus vehicula elit laoreet rhoncus. Vestibulum a laoreet
                neque, et blandit odio. Suspendisse a pretium massa. Cras lacinia est
                nisi, sollicitudin aliquam nisl tincidunt eu. Sed dignissim magna non
                est blandit, eu euismod diam laoreet. Pellentesque condimentum,
                mauris fermentum scelerisque elementum, sapien erat varius metus,
                vitae pretium mi metus dignissim turpis. Nullam nec sapien urna. Sed
                vulputate sagittis congue. Cras fringilla ex a massa convallis
                pretium. Ut eleifend sem sapien, a fringilla justo scelerisque eget.
            </Tabs.Tab>
        </Tabs>
    );
};
