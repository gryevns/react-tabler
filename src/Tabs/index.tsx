import React from "react";
import cn from "classnames";

const TabContext = React.createContext({
    active: undefined,
    setActive: (id: string): any => {},
});

export interface TabPanelProps {
    id: string;
    children: React.ReactNode;
    className?: string;
}

const TabPanel = ({ id, children, className }: TabPanelProps) => {
    const { active } = React.useContext(TabContext);
    const classes = cn(
        "tab-pane",
        { active: active === id, show: active === id },
        className,
    );
    return (
        <div className={classes} id={id}>
            {children}
        </div>
    );
};

export interface TabContentProps {
    children: React.ReactNode;
}

const TabContent = ({ children }: TabContentProps) => (
    <div className="tab-content">{children}</div>
);

export interface TabProps {
    id: string;
    children: React.ReactNode;
}

const Tab = ({ id, children }: TabProps) => {
    const { active, setActive } = React.useContext(TabContext);
    const classes = cn("nav-link", { active: active === id });
    const onClick = (e: React.MouseEvent) => {
        setActive(id);
        e.preventDefault();
    };
    return (
        <li className="nav-item">
            <a
                href={`#${id}`}
                className={classes}
                data-toggle="tab"
                onClick={onClick}
            >
                {children}
            </a>
        </li>
    );
};

export interface TabsHeaderProps {
    children: React.ReactNode;
    className?: string;
    fill?: boolean;
    alt?: boolean;
}

const TabsHeader = ({
    children,
    className,
    fill = false,
    alt = false,
}: TabsHeaderProps) => {
    const classes = cn(
        "nav nav-tabs",
        { "nav-fill": fill, "nav-tabs-alt": alt },
        className,
    );
    return (
        <ul className={classes} data-toggle="tabs">
            {children}
        </ul>
    );
};

export interface TabsProps {
    children: React.ReactNode;
    defaultEventKey: string;
}

const Tabs = ({ children, defaultEventKey }: TabsProps) => {
    const [active, setActive] = React.useState(defaultEventKey);
    return (
        <TabContext.Provider value={{ active, setActive }}>
            {children}
        </TabContext.Provider>
    );
};

Tabs.Tab = Tab;
Tabs.Content = TabContent;
Tabs.Panel = TabPanel;
Tabs.Header = TabsHeader;

export default Tabs;
