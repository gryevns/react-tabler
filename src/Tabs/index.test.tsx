import React from "react";
import { mount } from "enzyme";

import Tabs from "./index";

const Test = props => (
    <Tabs {...props}>
        <Tabs.Tab title="c1" eventKey="c1">
            content1
        </Tabs.Tab>
        <Tabs.Tab title="c2" eventKey="c2">
            content2
        </Tabs.Tab>
    </Tabs>
);

describe("Tabs", () => {
    it("should match snapshot", () => {
        const wrapper = mount(<Test />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match cards snapshot", () => {
        const wrapper = mount(<Test cards />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match fill snapshot", () => {
        const wrapper = mount(<Test fill />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should match alt snapshot", () => {
        const wrapper = mount(<Test alt />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should load first tab by default", () => {
        const wrapper = mount(<Test />);
        expect(wrapper.text().includes("content1")).toBeTruthy();
        expect(wrapper.text().includes("content2")).toBeFalsy();
    });
    it("should load defaultEventKey tab by default", () => {
        const wrapper = mount(<Test defaultEventKey="c2" />);
        expect(wrapper.text().includes("content1")).toBeFalsy();
        expect(wrapper.text().includes("content2")).toBeTruthy();
    });
    it("should switch tab by click", () => {
        const wrapper = mount(<Test defaultEventKey="c1" />);
        expect(wrapper.text().includes("content1")).toBeTruthy();
        wrapper
            .find("li")
            .at(1)
            .simulate("click");
        expect(wrapper.text().includes("content2")).toBeTruthy();
    });
});
