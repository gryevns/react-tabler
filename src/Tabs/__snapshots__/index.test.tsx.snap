// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match alt snapshot 1`] = `
<Test
  alt={true}
>
  <Tabs
    alt={true}
  >
    <DefaultTabs
      alt={true}
      tabs={
        Object {
          "c1": Object {
            "children": "content1",
            "title": "c1",
          },
          "c2": Object {
            "children": "content2",
            "title": "c2",
          },
        }
      }
    >
      <Card>
        <div
          className="card"
        >
          <Nav
            alt={true}
            tabs={
              Object {
                "c1": Object {
                  "children": "content1",
                  "title": "c1",
                },
                "c2": Object {
                  "children": "content2",
                  "title": "c2",
                },
              }
            }
          >
            <ul
              className="nav nav-tabs nav-tabs-alt"
            >
              <li
                className="nav-item"
                key="tab-c1"
                onClick={[Function]}
              >
                <a
                  className="nav-link active"
                >
                  c1
                </a>
              </li>
              <li
                className="nav-item"
                key="tab-c2"
                onClick={[Function]}
              >
                <a
                  className="nav-link"
                >
                  c2
                </a>
              </li>
            </ul>
          </Nav>
          <CardBody>
            <div
              className="card-body"
            >
              <div
                className="tab-content"
              >
                <div
                  className="tab-pane active"
                >
                  content1
                </div>
              </div>
            </div>
          </CardBody>
        </div>
      </Card>
    </DefaultTabs>
  </Tabs>
</Test>
`;

exports[`Tabs should match cards snapshot 1`] = `
<Test
  cards={true}
>
  <Tabs
    cards={true}
  >
    <CardTabs
      tabs={
        Object {
          "c1": Object {
            "children": "content1",
            "title": "c1",
          },
          "c2": Object {
            "children": "content2",
            "title": "c2",
          },
        }
      }
    >
      <div
        className="card-tabs"
      >
        <Nav
          tabs={
            Object {
              "c1": Object {
                "children": "content1",
                "title": "c1",
              },
              "c2": Object {
                "children": "content2",
                "title": "c2",
              },
            }
          }
        >
          <ul
            className="nav nav-tabs"
          >
            <li
              className="nav-item"
              key="tab-c1"
              onClick={[Function]}
            >
              <a
                className="nav-link active"
              >
                c1
              </a>
            </li>
            <li
              className="nav-item"
              key="tab-c2"
              onClick={[Function]}
            >
              <a
                className="nav-link"
              >
                c2
              </a>
            </li>
          </ul>
        </Nav>
        <div
          className="tab-content"
        >
          <div
            className="tab-pane card active"
          >
            <CardBody>
              <div
                className="card-body"
              >
                content1
              </div>
            </CardBody>
          </div>
        </div>
      </div>
    </CardTabs>
  </Tabs>
</Test>
`;

exports[`Tabs should match fill snapshot 1`] = `
<Test
  fill={true}
>
  <Tabs
    fill={true}
  >
    <DefaultTabs
      fill={true}
      tabs={
        Object {
          "c1": Object {
            "children": "content1",
            "title": "c1",
          },
          "c2": Object {
            "children": "content2",
            "title": "c2",
          },
        }
      }
    >
      <Card>
        <div
          className="card"
        >
          <Nav
            fill={true}
            tabs={
              Object {
                "c1": Object {
                  "children": "content1",
                  "title": "c1",
                },
                "c2": Object {
                  "children": "content2",
                  "title": "c2",
                },
              }
            }
          >
            <ul
              className="nav nav-tabs nav-fill"
            >
              <li
                className="nav-item"
                key="tab-c1"
                onClick={[Function]}
              >
                <a
                  className="nav-link active"
                >
                  c1
                </a>
              </li>
              <li
                className="nav-item"
                key="tab-c2"
                onClick={[Function]}
              >
                <a
                  className="nav-link"
                >
                  c2
                </a>
              </li>
            </ul>
          </Nav>
          <CardBody>
            <div
              className="card-body"
            >
              <div
                className="tab-content"
              >
                <div
                  className="tab-pane active"
                >
                  content1
                </div>
              </div>
            </div>
          </CardBody>
        </div>
      </Card>
    </DefaultTabs>
  </Tabs>
</Test>
`;

exports[`Tabs should match snapshot 1`] = `
<Test>
  <Tabs>
    <DefaultTabs
      tabs={
        Object {
          "c1": Object {
            "children": "content1",
            "title": "c1",
          },
          "c2": Object {
            "children": "content2",
            "title": "c2",
          },
        }
      }
    >
      <Card>
        <div
          className="card"
        >
          <Nav
            tabs={
              Object {
                "c1": Object {
                  "children": "content1",
                  "title": "c1",
                },
                "c2": Object {
                  "children": "content2",
                  "title": "c2",
                },
              }
            }
          >
            <ul
              className="nav nav-tabs"
            >
              <li
                className="nav-item"
                key="tab-c1"
                onClick={[Function]}
              >
                <a
                  className="nav-link active"
                >
                  c1
                </a>
              </li>
              <li
                className="nav-item"
                key="tab-c2"
                onClick={[Function]}
              >
                <a
                  className="nav-link"
                >
                  c2
                </a>
              </li>
            </ul>
          </Nav>
          <CardBody>
            <div
              className="card-body"
            >
              <div
                className="tab-content"
              >
                <div
                  className="tab-pane active"
                >
                  content1
                </div>
              </div>
            </div>
          </CardBody>
        </div>
      </Card>
    </DefaultTabs>
  </Tabs>
</Test>
`;
